### PR TITLE
rework event handling to allow sending data to `Control`

### PR DESCRIPTION
**Pull request by kbleeke**
This expands the event handling previously used for the join ioctls.

- Adds an enum for the event status field ([from here](https://github.com/Infineon/wifi-host-driver/blob/master/WiFi_Host_Driver/src/include/whd_events_int.h#L195))
- Adds an event publishing can be enabled and disabled by `Control`
- Adds Payload enum to carry event data. Currently empty. Will be used for wifi scanning.

### CODE REVIEW

File: src/events.rs [hunk 1]
```diff
@@ -284,13 +284,114 @@ pub enum Event {
...
+impl Message {
+    pub fn new(status: Status, payload: Payload) -> Self {
+        Self {
+            header: status,
+            payload,
+        }
+    }
+}
+
+const EVENT_BITS: usize = ((Event::LAST as usize + 31) & !31) / 32;
```
Comment (Dirbaio) [inline src/events.rs:334]: this is equivalent to just `(Event::LAST as usize + 31) / 32`

File: src/events.rs [hunk 2]
```diff
@@ -284,13 +284,114 @@ pub enum Event {
...
+
+#[derive(Default)]
+struct EventMask {
+    mask: [u32; EVENT_BITS],
+}
+
+impl EventMask {
+    fn enable(&mut self, event: Event) {
+        let n = event as u32;
+        let word = n >> 5;
```
Comment (Dirbaio) [inline src/events.rs:344]: took me a while to realize this is correct. IMO `n/32` and `n%32` would be clearer. The compiler can optimize these to bitwise stuff.

**Comment by Dirbaio** [comment]
I'm not sure if a pubsub is the way to go for managing events. It seems inherently racy:

- Sending uses `publish_immediate`, which loses events if the queue is full. It's possible they'll get lost if they arrive too fast. Using `publish` instead is not a fix, it could deadlock if the control task is waiting for the bus task to process an ioctl, and the bus task is waiting for the control task to pop some events to make space in the pubsub.
- You start a join, the event arrives, then the event task cancels the join before the event is popped. The event is left in the queue, the next join will wrongly complete "instantly" because it'll see the stale event.
- The join could complete very fast, before `wait_for_join()` gets called and enables the mask, causing the event to be lost.

I wonder if it's better to handle all events synchronously in the bus task instead, updating some shared state. For example for join it'd set `join_done = true` in the shared state, and `wait_for_join()` would wait for that.

**Comment by kbleeke** [comment]
In general, I think some way to communicate events to `Control` would be necessary if a user of this library wants to process scan results. 

>Sending uses publish_immediate, which loses events if the queue is full. 

True but I figured it would still work in this case. As far as I can tell, the task must yield before the next event can be received (it will certainly yield when trying to read from the bus for the next event). This gives event consumers a chance to process for event. Of course, this restricts what the consumer can do without missing events. This would be a concern consuming scan results.

>reading stale events or losing events

My understanding is that this isn't possible. I create a new subscriber for each join ioctl that should only receive events that happen after its creation (according to docs). I assumed here that nothing happens before the SET_SSID ioctl and the subscriber and event mask are set before this ioctl is issued. 



Regarding using publish:

Would it work, if I make sure that no subscriber exists when issuing an ioctl? Assuming that dropping a subscriber releasess the subscriber's unread items and a pubsub queue with no subscribers is never full?

**Comment by kbleeke** [comment]
Regarding deadlocking the task:

I think it already has a similar issue [here](https://github.com/embassy-rs/cyw43/blob/master/src/runner.rs#L420) where the problem is solved by simply dropping the incoming data. 

In principle, the potential deadlock here could by fixed by splitting the stack into a tx and an rx future that run concurrently?

**Comment by Dirbaio** [comment]
Yes, I agree that with an extra dose of care these problems can be mitigated. It feels a bit off to me that these problems are possible at all though, while a different approach could prevent them at the root.

I'm fine with carrying forward with your approach though. We can always refactor later if it turns out to be troublesome. And it's you who's doing all the work after all :) 

Could you check the other comments?

**Comment by kbleeke** [comment]
I left some comments regarding the event handling in the code.

**Comment by Dirbaio** [comment]
:rocket: Thank you!

**Pull request merged**